core/exec.c

```diff
@@ -229,6 +229,171 @@ static void lai_exec_reduce_op(int opcode, lai_state_t *state, struct lai_operan
 		lai_operand_mutate(state, &operands[1], &result);
 		break;
 	}
+	case CONCAT_OP:
+	{
+		LAI_CLEANUP_VAR lai_variable_t operand0 = LAI_VAR_INITIALIZER;
+		lai_exec_get_objectref(state, &operands[0], &operand0);
+		LAI_CLEANUP_VAR lai_variable_t operand1 = LAI_VAR_INITIALIZER;
+		lai_exec_get_objectref(state, &operands[1], &operand1);
+
+		LAI_CLEANUP_VAR lai_variable_t operand0_convert = LAI_VAR_INITIALIZER;
+		LAI_CLEANUP_VAR lai_variable_t operand1_convert_temp = LAI_VAR_INITIALIZER;
+		LAI_CLEANUP_VAR lai_variable_t operand1_convert = LAI_VAR_INITIALIZER;
+
+		/*
+		 * Convert object types other than integers, strings and buffers to a string.
+		 *
+		 * According to the ACPI specification buffer fields should be turned into
+		 * strings, but in ACPICA they are treated as either integers on strings
+		 * based on their size.
+		 */
+		if(operand0.type != LAI_INTEGER && operand0.type != LAI_BUFFER && operand0.type != LAI_STRING) {
+			if(operand0.type == LAI_HANDLE) {
+				lai_api_error_t error = lai_create_string(&operand0_convert, 0);
+				if(error != LAI_ERROR_NONE) {
+					lai_panic("failed to allocate memory for AML string");
+				}
+				error = lai_obj_to_type_string(&operand0_convert, operand0.handle);
```

Comment:
Instead of allocating a 0-sized string and then reallocating it in `lai_obj_to_type_string`, you could just make `lai_obj_to_type_string` create the correctly sized string.